### PR TITLE
Minting data types and initial LedgerDb support

### PR DESCRIPTION
**Pull request by eranrund**
### Motivation

As part of the minting proof of concept work we know that we need the following:
* A data structure to represent the currently active minting configurations - this controls who is currently allowed to mint which tokens, and at what limit
* This configuration needs to be stored inside the ledger, so that at any point we can have consensus query the configuration for a given token id, as well as update the total amount minted via a given configuration. This will be used by the new minting transaction validation code and the updated block forming logic.
* New SCP transactions for minting and updating the minting configuration. 

### In this PR
This PR contains some of the initial building blocks for this:
* New data structures to represent a multisig-signed minting transaction, multisig-signed set-minting-config transaction and other associated data structures
* API changes to `LedgerDB` that allow us to store the list of currently active minting configurations per token id, the list of already-seen nonces, and the total amount minted by a specific configuration.

### Future Work
* Include the new transaction types inside `BlockContents` and modify `append_block()` to propagate new minting configurations into the new `MintConfigStore`
* Transaction validation code for the new transaction types
* SCP changes to support the new transaction types
* Enclave block forming changes to support minting
* Tests
* Probably many other things...

### CODE REVIEW

**Review by remoun — APPROVED** [review]

File: consensus/service/src/config/tokens.rs [hunk 1]
```diff
@@ -5,12 +5,74 @@
...
+                        hex::decode(s)
+                            .map_err(serde::de::Error::custom)
+                            .and_then(|bytes| {
+                                // Decode each byte vector into an Ed25519Public.
+                                Ed25519Public::try_from_der(&bytes[..])
+                                    .map_err(serde::de::Error::custom)
+                            })
+                    })
+                    // Return the keys.
+                    .collect::<Result<_, D::Error>>()?;
```
[the remark this answers is not in the document]
Comment (eranrund) [reply consensus/service/src/config/tokens.rs:66]: The compiler is not smart enough to figure that on its own:
```
error[E0284]: type annotations needed
  --> consensus/service/src/config/tokens.rs:66:22
   |
66 |                     .collect()?;
   |                      ^^^^^^^ cannot infer type
```

File: ledger/migration/src/lib.rs [hunk 2]
```diff
@@ -80,6 +80,26 @@ pub fn migrate(ledger_db_path: impl AsRef<Path>, logger: &Logger) {
...
+            Err(MetadataStoreError::VersionIncompatible(20200707, _)) => {
+                log::info!(
+                    logger,
+                    "Ledger db migrating from version 20200707 to 20220222..."
+                );
+                MintConfigStore::create(&env).expect("Failed creating MintConfigStore");
+
+                let mut db_txn = env.begin_rw_txn().expect("Failed starting rw transaction");
+                metadata_store
+                    .set_version_to_latest(&mut db_txn)
```
Comment (wjuan-mob) [inline ledger/migration/src/lib.rs:93]: Sorry for my ignorance, but is there a benefit of setting to latest vs specifying 20220222 here?

File: ledger/db/src/tx_out_store.rs [hunk 3]
```diff
@@ -796,7 +796,7 @@ pub mod tx_out_store_tests {
     use std::path::Path;
     use tempdir::TempDir;
 
-    fn get_env() -> Environment {
+    pub fn get_env() -> Environment {
```
Comment (wjuan-mob) [inline ledger/db/src/tx_out_store.rs:799]: If we're making this a public function, maybe we should add a doc comment? Or is it not necessary here?

File: ledger/db/src/lib.rs [hunk 4]
```diff
@@ -45,10 +47,11 @@ use std::{
 pub use error::Error;
 pub use ledger_trait::{Ledger, MockLedger};
 pub use mc_util_lmdb::{MetadataStore, MetadataStoreError};
+pub use mint_config_store::{ActiveMintConfig, MintConfigStore};
 pub use tx_out_store::TxOutStore;
 
 pub const MAX_LMDB_FILE_SIZE: usize = 2usize.pow(40); // 1 TB
-pub const MAX_LMDB_DATABASES: u32 = 22; // maximum number of databases in the lmdb file
+pub const MAX_LMDB_DATABASES: u32 = 24; // maximum number of databases in the lmdb file
```
Comment (wjuan-mob) [inline ledger/db/src/lib.rs:54]: Maybe it would help to note somewhere in the PR body that we are adding 2 DBs? A mapping of token id -> currently active mint configurations and a mapping of nonce -> SetMintConfigTx?

File: ledger/db/src/mint_config_store.rs [hunk 5]
```diff
@@ -0,0 +1,573 @@
...
+
+//! Data access abstraction for minting configuration stored in the ledger.
+//!
+//! This store maintains two LMDB databases:
+//! 1) A mapping of token id -> currently active mint configurations.
+//!    This database is used for two things:
+//!      1) It allows transaction validation code to figure out if a mint
+//! transaction is allowed to mint.
+//!      2) It enables keeping track of how much was minted using a given
+//! configuration. This is used to enforce the per-configuration mint limit. 2)
```
Comment (wjuan-mob) [inline ledger/db/src/mint_config_store.rs:11]: Nit: I think this 2) should start on the next line.

File: ledger/db/src/mint_config_store.rs [hunk 6]
```diff
@@ -0,0 +1,573 @@
...
+        )?;
+        env.create_db(
+            Some(SET_MINT_CONFIG_TX_BY_NONCE_DB_NAME),
+            DatabaseFlags::empty(),
+        )?;
+        Ok(())
+    }
+
+    /// Set mint configurations for a given token.
+    pub fn set_active_mint_configs(
```
Comment (wjuan-mob) [inline ledger/db/src/mint_config_store.rs:98]: Sorry I was a little confused by the intention of this. Could we add a test that demonstrates the expected behaviour when we deactivate a mint config via this method or reactivate a mint config via this method?

File: ledger/db/src/lib.rs [hunk 7]
```diff
@@ -738,6 +774,15 @@ pub fn key_bytes_to_u64(bytes: &[u8]) -> u64 {
     u64::from_be_bytes(bytes.try_into().unwrap())
 }
 
+pub fn u32_to_key_bytes(value: u32) -> [u8; 4] {
+    value.to_be_bytes()
+}
+
+pub fn key_bytes_to_u32(bytes: &[u8]) -> u32 {
```
Comment (remoun) [inline ledger/db/src/lib.rs:781]: I don't see where this method is used.

File: ledger/db/src/lib.rs [hunk 8]
```diff
@@ -738,6 +774,15 @@ pub fn key_bytes_to_u64(bytes: &[u8]) -> u64 {
...
+    assert_eq!(8, bytes.len());
```
Comment (remoun) [inline ledger/db/src/lib.rs:782]: Should this be 4?

File: consensus/service/src/config/tokens.rs [hunk 9]
```diff
@@ -447,7 +574,7 @@ mod tests {
         assert_eq!(tokens, tokens2);
 
         // Validation should fail since allow_any_fee cannot be used on non-MOB tokens.
-        assert!(tokens.validate().is_err());
+        assert_validation_error(&tokens, "MOB token configuration not found");
```
Comment (wjuan-mob) [inline consensus/service/src/config/tokens.rs:577]: Nit: Shouldn't this be: allow_any_fee can only be used for MOB

**Review by wjuan-mob — APPROVED** [review]

File: consensus/service/src/config/tokens.rs [hunk 10]
```diff
@@ -5,12 +5,74 @@
...
 use std::{fs, iter::FromIterator, path::Path};
 
+mod der_signer_set {
+    use super::*;
+
+    /// A helper struct for ser/derserializing a SignerSet that uses hex-encoded
+    /// DER representation of Ed25519 public keys.
+    #[derive(Serialize, Deserialize)]
+    struct DerSignerSet {
+        signers: Vec<String>,
```
Comment (jcape) [inline consensus/service/src/config/tokens.rs:21]: DER is the binary format, and that gets Base64 encoded for PEM, and elsewhere I've used `Vec<Vec<u8>>` for this...

Comment (jcape) [reply consensus/service/src/config/tokens.rs:21]: c.f. https://docs.rs/pem/latest/pem/fn.parse_many.html

File: ledger/migration/src/lib.rs (hunk 2 above)
Comment (eranrund) [reply ledger/migration/src/lib.rs:93]: Good catch, that was actually a potential bug!
We need to be explicit with the version we set to (and not just set to latest) so that incremental migration steps would work.

File: ledger/db/src/tx_out_store.rs (hunk 3 above)
Comment (eranrund) [reply ledger/db/src/tx_out_store.rs:799]: I'll add a comment, but in general I don't think we made it a priority to document test code thoroughly.

File: ledger/db/src/lib.rs (hunk 4 above)
Comment (eranrund) [reply ledger/db/src/lib.rs:54]: I think that becomes clear when reading the code, not sure where I would note this / who would ever look at it after it gets merged.

File: ledger/db/src/mint_config_store.rs (hunk 6 above)
Comment (eranrund) [reply ledger/db/src/mint_config_store.rs:98]: I believe this is covered by `set_get_behaves_correctly`

File: consensus/service/src/config/tokens.rs (hunk 9 above)
Comment (eranrund) [reply consensus/service/src/config/tokens.rs:577]: good catch

File: ledger/db/src/mint_config_store.rs (hunk 6 above)
Comment (wjuan-mob) [reply ledger/db/src/mint_config_store.rs:98]: I believe that set_get_behaves_correctly is testing setting the mint config and retrieving that same config for 2 different tokens. What I'm unsure about is the behaviour of deactivating an active config on a token, e.g. we go from having an active mint config to an empty active mint config, perhaps after having already minted some tokens. Then reactivating the previously deactivated config.

Comment (eranrund) [reply ledger/db/src/mint_config_store.rs:98]: Minting configurations can be replaced at any time and do not affect anything previously minted, they just control who can currently mint. There is no deactivation - other than by replacing the currently active configuration with an empty one`set_replaces_configuration` checks that an existing configuration can be replaced by a new one. I'll add a test that verifies an empty one is accepted.

File: consensus/service/src/config/tokens.rs [hunk 11]
```diff
@@ -521,4 +655,176 @@ mod tests {
             Some(1)
         );
     }
+
+    #[test]
+    fn master_minters_deserialize_serialize_deserialize_works() {
```
Comment (remoun) [inline consensus/service/src/config/tokens.rs:660]: Nit: This test only does one ser-der round trip, so remove the first `deserialize_`

File: ledger/db/src/ledger_trait.rs [hunk 12]
```diff
@@ -90,4 +91,16 @@ pub trait Ledger: Send {
...
+    /// set-mint-config transaction.
+    fn set_active_mint_configs(&self, set_mint_config_tx: &SetMintConfigTx) -> Result<(), Error>;
+
+    /// Get active mint configurations for a given token id.
+    /// Returns an empty array of no mint configurations are active for the
+    /// given token id.
+    fn get_active_mint_configs(&self, token_id: TokenId) -> Result<Vec<ActiveMintConfig>, Error>;
+
+    /// Update the total minted amount for a given MintConfig.
+    fn update_total_minted(&self, mint_config: &MintConfig, amount: u64) -> Result<(), Error>;
```
Comment (remoun) [inline ledger/db/src/ledger_trait.rs:105]: This seems racy if there are multiple active minters. I think it would be more robust as a `log_minted` kind of method that increments the `total_minted`, which would be more retryable.

IIUC we're only planning on having one minter, so this might not be an issue in practice, but also the `log_minted` approach could enforce that the total is monotonically increasing.

**Review by remoun — CHANGES_REQUESTED** [review]

File: ledger/db/src/ledger_trait.rs (hunk 12 above)
Comment (eranrund) [reply ledger/db/src/ledger_trait.rs:105]: I don't believe this is racy since this the read-modify-write is done inside a single LMDB transaction.

Comment (eranrund) [reply ledger/db/src/ledger_trait.rs:105]: Just to clarify, the total minted amount is tracked per minting configuration. So there could be multiple `MintConfig`s currently configured, but the total minted amount is tracked individually for each one.

A followup PR might actually remove this from the public interface, since it can be done inside `append_block` - as the block has information about who minted what so there will not be a need for an external caller to call this.

File: consensus/service/src/config/tokens.rs (hunk 10 above)
Comment (eranrund) [reply consensus/service/src/config/tokens.rs]: Done

File: consensus/service/src/config/tokens.rs [hunk 13]
```diff
@@ -11,61 +11,63 @@ use mc_transaction_core::{tokens::Mob, Token, TokenId};
...
-            signers: signer_set
+        let pem_signer_set = signer_set.as_ref().map(|signer_set| {
+            let pems = signer_set
                 .signers()
                 .iter()
-                .map(|signer| hex::encode(&signer.to_der()))
-                .collect(),
-            threshold: signer_set.threshold(),
+                .map(|signer| Pem {
+                    tag: String::from("FOO"),
```
Comment (remoun) [inline consensus/service/src/config/tokens.rs:36]: How about "signer"?
```suggestion
                    tag: "signer".to_string(),
```

File: consensus/service/src/config/tokens.rs [hunk 14]
```diff
@@ -11,61 +11,63 @@ use mc_transaction_core::{tokens::Mob, Token, TokenId};
...
+                let pems = pem::parse_many(pem_signer_set.signers.as_bytes())
+                    .map_err(serde::de::Error::custom)?;
+
+                let signers = pems
                     .iter()
-                    .map(|s| {
-                        // Decode each hex-encoded string into a byte vector.
-                        hex::decode(s)
+                    .map(|pem| {
+                        Ed25519Public::try_from_der(&pem.contents[..])
```
Comment (remoun) [inline consensus/service/src/config/tokens.rs:64]: Would it make sense to check the tag here?

File: consensus/service/src/config/tokens.rs (hunk 13 above)
Comment (eranrund) [reply consensus/service/src/config/tokens.rs:36]: oops, I missed that. it should be `PUBLIC KEY` to match what openssl generates, but in general has no meaningful implication.

File: consensus/service/src/config/tokens.rs (hunk 14 above)
Comment (eranrund) [reply consensus/service/src/config/tokens.rs:64]: I don't think there's a point, the tag doesn't serve any meaningful purpose. Any tag would be fine and will depend on what the user used to encode the PEM.

File: ledger/db/src/mint_config_store.rs [hunk 15]
```diff
@@ -0,0 +1,631 @@
...
+            self.get_active_mint_configs(TokenId::from(mint_config.token_id), db_transaction)?;
+
+        // Find the active mint config that matches the mint config we were given.
+        let active_mint_config = active_mint_configs
+            .iter_mut()
+            .find(|active_mint_config| active_mint_config.mint_config == *mint_config)
+            .ok_or_else(|| Error::InvalidMintConfig("Mint config not found".to_string()))?;
+
+        // Update the total minted amount.
+        active_mint_config.total_minted = amount;
```
Comment (remoun) [inline ledger/db/src/mint_config_store.rs:176]: Consider enforcing that `amount > total_minted` here, and adding a test that exercises it.

**Review by remoun — APPROVED** [review]

**Review by jcape — APPROVED** [review]

File: ledger/db/src/lib.rs [hunk 16]
```diff
@@ -76,7 +79,7 @@ impl MetadataStoreSettings for LedgerDbMetadataStoreSettings {
     // db opening for any incompatibilities, and either refuse to open or
     // perform a migration.
     #[allow(clippy::unreadable_literal)]
-    const LATEST_VERSION: u64 = 20200707;
+    const LATEST_VERSION: u64 = 20220222;
```
[the remark this answers is not in the document]
Comment (eranrund) [reply ledger/db/src/lib.rs:82]: Nice

File: consensus/service/src/config/tokens.rs [hunk 17]
```diff
@@ -689,25 +712,26 @@ mod tests {
...
-                "302a300506032b65700321007b7b7b7b7b7b7b7b7b7b7b7b7b7b7b7b7b7b7b7b7b7b7b7b7b7b7b7b7b7b7b7b"
-            ]
+            signers = """
+            -----BEGIN PUBLIC KEY-----
+            MCowBQYDK2VwAyEAyj6m0NRTlw/R28Q+R7vBakwybuaNFneKrvRVAYNp5WQ=
+            -----END PUBLIC KEY-----
+            -----BEGIN PUBLIC KEY-----
+            MCowBQYDK2VwAyEAl3XVo/DeiTjHn8dYQuEtBjQrEWNQSKpfzw3X9dewSVY=
+            -----END PUBLIC KEY-----
+            """
```
Comment (remoun) [inline consensus/service/src/config/tokens.rs:722]: Optional/Equivalent: These keys seem to repeat across test cases, so how about we use consts for them? e.g.

```rust
let input = format!(r#"
  ...
  signers="""
    {}
    {}
   """
"#,
  PEM_ENCODED_KEY_1,
  PEM_ENCODED_KEY_2,
);
```

**Review by remoun — APPROVED** [review]

File: consensus/service/src/config/tokens.rs (hunk 17 above)
Comment (eranrund) [reply consensus/service/src/config/tokens.rs:722]: My personal opinion is that I think its fairly minimal and easier to read without the string formatting. Part of what I like in having the full TOML/JSON is that it serves as an easy-to-read example of what the file should look like.

**Review by mfaulk — APPROVED** [review]
